oak_restricted_kernel/src/interface.rs

```diff
@@ -0,0 +1,94 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! Exported interface for interacting with the kernel from C.
+
+use alloc::alloc::{alloc, dealloc, Layout};
+use core::{
+    cell::{LazyCell, RefCell},
+    ffi::c_void,
+    ptr::null_mut,
+};
+use hashbrown::HashMap;
+use spinning_top::{const_spinlock, Spinlock};
+use static_assertions::const_assert;
+
+/// The maximum alignment that we need to support any C type that has fundamental alignment.
+const MAX_ALIGNMENT: usize = core::mem::align_of::<libc::max_align_t>();
+
+const_assert!(MAX_ALIGNMENT.is_power_of_two());
+
+/// Cache for keeping track of which memory locations we allocated and the layout (size and
+/// alignment) that was used for each allocation.
+static ALLOCATED_ITEMS: Spinlock<LazyCell<RefCell<HashMap<usize, Layout>>>> =
+    const_spinlock(LazyCell::new(|| RefCell::new(HashMap::new())));
+
+/// Allocates memory of the given size with alignment that is suitable for any C type that has
+/// fundamental alignment.
+///
+/// This function returns a pointer to the allocated memory if the allocation succeeds and null if
+/// it fails. All allocated memory must be deallocated using `heap_free`.
+#[no_mangle]
+pub extern "C" fn heap_alloc(size: usize) -> *mut c_void {
```

Comment:
Is this not called just `alloc` because of name clashes? If so, I would add a comment about that, since people in the future may wonder about that and be tempted to rename it.

Reply:
Yes (not called `malloc` for this reason. Added a comment here and for `heap_free`.